### DISCUSSION TITLE
Mimic BlockingChannel.basic_ack for asyncio.

### DISCUSSION THREAD

**Discussion by NiekHoekstra** [pika/pika#1502] [Q&A] [answered]
I'm trying to mimic the behavior of the `BlockingChannel.basic_ack` in asyncio.
I'm using a `Future` object and set the result (or exception).

The callback for `add_on_close_callback` works like a charm for errors, I just call `Future.set_exception` and it's fine.
What I'm not sure about is how to "verify" if the `ack` was correct.

I tried looking at the `BlockingConnection`, which (I think) uses its transport layer to check if transmit buffer is empty.
To do that in asyncio seems messy, or I'm just not looking in the right spot.

At the moment, I'm abusing  `sleep` and assume success after a while (having given ample opportunity for `close` callbacks).
The `asyncio.sleep` feels a bit weird in my head.
Any advice or recommendations to do this cleaner?

The following code is a bare example, with a `database` context thrown in.
The idea is that the database shouldn't `commit()` if the `ack` to RMQ fails.

```python
import asyncio
from functools import partial
from typing import Optional

from pika.channel import Channel
from pika.spec import Basic, BasicProperties


class Transaction:
    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if exc_val is not None:
            await self.rollback()

    async def commit(self):
        print('committed')

    async def rollback(self):
        print('rollback')


class Example:
    def __init__(self, channel: Channel, queue: str = 'test-queue'):
        self.channel: Channel = channel
        self.channel.add_on_close_callback(self.on_close)
        self.channel.basic_consume(queue, on_message_callback=self.on_message)
        self.loop = asyncio.get_event_loop()
        self._pending: Optional[asyncio.Future] = None

    def on_message(self, ch: Channel, basic: Basic.Deliver, props: BasicProperties = None,
                   data: bytes = None):
        coro = self.process(ch, basic, props, data)
        task = self.loop.create_task(coro)
        self._pending = self.loop.create_future()

    async def process(self, ch: Channel, basic: Basic.Deliver, props: BasicProperties, data: bytes):
        async with Transaction() as database_transaction:
            ch.basic_ack(basic.delivery_tag, multiple=False)
            task = self._pending._loop.create_task(asyncio.sleep(0.2))
            task.add_done_callback(partial(self.on_timeout, self._pending))
            await self._pending
            await database_transaction.commit()

    def on_timeout(self, future: asyncio.Future, task: asyncio.Future):
        if task.cancelled():
            return
        if future.done():
            return
        print('confirming')
        if not future.done():
            future.set_result(None)

    def on_close(self, channel: Channel, reason: Exception):
        if self._pending and not self._pending.done():
            self._pending.set_exception(reason)
```

**Comment by lukebakken** [comment, marked as answer]
> The idea is that the database shouldn't commit() if the ack to RMQ fails

An alternative would be to not care about the ack. An ack will only fail if there is a network or operating system issue where the data can't be sent to RabbitMQ. Once the data leaves the computer running your application you can't ever really be sure it made it to RabbitMQ and was processed, because RabbitMQ doesn't ack an ack 😉 

Your code should (must) handle the rare case where a message is redelivered because RabbitMQ did not receive the ack.

**Reply by NiekHoekstra** [reply]
I was afraid you might say that :joy:

You're absolutely right. 
I'm trying to fix up a legacy project, which doesn't quite use RMQ as it should.

Seems like I'll have to redesign a bit of it.
Being able to deal with redelivered messages solves multiple (potential) issues.
Adding those internal checks and balances is a good idea.

Thanks for your time, much appreciated.
